lib/logstash/outputs/elastic_app_search.rb

```diff
@@ -64,17 +67,32 @@ def format_batch(events)
         doc["id"] = event.sprintf(@document_id)
       end
       doc.delete("@version")
-      doc
+      resolved_engine = event.sprintf(@engine)
+      unless docs_for_engine[resolved_engine]
+        if @logger.trace?
+          @logger.trace("Creating new engine segment in batch to send", :resolved_engine => resolved_engine)
+        end
+        docs_for_engine[resolved_engine] = []
+      end
+      docs_for_engine[resolved_engine] << doc
     end
+    docs_for_engine
   end
 
-  def index(documents)
-    response = @client.index_documents(@engine, documents)
-    report(documents, response)
-  rescue => e
-    @logger.error("Failed to execute index operation. Retrying..", :exception => e.class, :reason => e.message)
-    sleep(1)
-    retry
+  def index(batch)
+    batch.each do |resolved_engine, documents|
+      begin
+        if resolved_engine =~ ENGINE_WITH_SPRINTF_REGEX || resolved_engine =~ /^\s*$/
+          raise "Engine field name #{@engine} doesn't exists or resolves to empty string"
+        end
+        response = @client.index_documents(resolved_engine, documents)
+        report(documents, response)
+      rescue => e
+        @logger.error("Failed to execute index operation. Retrying..", :exception => e.class, :reason => e.message)
```

Comment:
Maybe add more contextual information here - the `resolved_engine` value could be useful here, as the output from AppSearch does not contain the engine name in its 404 response.